File: 01.DFS递归/leetcode_90_subsetsWithDup.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
private:
    int n;
    vector<int> tmp;
    vector<int> N;
    vector<vector<int>> res;
    void dfs(int i){
        if (i == n){
            res.push_back(tmp);
            return;
        }

        int x = N[i];
        tmp.push_back(x);
        dfs(i+1);
        tmp.pop_back();

        i++;
        while (i<n && N[i] == x){
            i++;
        }
        dfs(i);
    }
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        n = nums.size();
        sort(nums.begin(), nums.end());
        N = nums;
        dfs(0);
        return res;
    }
};

int main(){
    Solution s;
    vector<int> nums = {4,4,4,1,4};
    vector<vector<int>> res = s.subsetsWithDup(nums);
    for (auto& i : res){
        for (auto& j : i){
            cout << j << " ";
        }
        cout << endl;
    }
}

File: 01.DFS递归/leetcode_90_subsetsWithDup.py
from typing import List
from collections import defaultdict

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nums.sort()
        res = []
        tmp = []
        # 定义dfs(i)，返回值为从第i个位置开始选，所有子集的情况，因此答案为dfs(0)
        def dfs(i):
            # 边界条件
            if i == n:
                return [tmp.copy()]

            # 选nums[i]
            ans = []
            x = nums[i]
            tmp.append(x)
            ans += dfs(i+1)
            tmp.pop()

            # 不选nums[i]
            i += 1
            while i<n and nums[i] == x:
                # 跳过重复元素，避免出现重复子集
                i += 1
            ans += dfs(i)
            return ans
        return dfs(0)
        # -------------------------------------------------------
        # res = defaultdict(int)
        # nums.sort()
        # for mode in range(0, 2<<len(nums)):
        #     tmp = []
        #     for i in range(len(nums)):
        #         if (mode>>i)&1:
        #             tmp.append(nums[i])
        #     res[tuple(tmp)] = 1
        # return list(map(list, list(res.keys())))

if __name__ == "__main__":
    s = Solution()
    nums = [4,4,4,1,4]
    print(s.subsetsWithDup(nums))
